tracer/src/Datadog.Trace/TraceContext.cs

```diff
@@ -100,30 +100,15 @@ public void AddSpan(Span span)
             {
                 if (RootSpan == null)
                 {
-                    // first span added is the root span
+                    // first span added is the local root span
                     RootSpan = span;
 
+                    // if we don't have a sampling priority yet, make a sampling decision now
                     if (_samplingPriority == null)
                     {
-                        if (span.Context.Parent is SpanContext { SamplingPriority: { } samplingPriority })
-                        {
```

Comment:
I don't follow why this code is incorrect now?

Reply:
In the first branch we check that the parent SpanContext has a SamplingPriority and if so we set the TraceContext's SamplingPriority accordingly. Tracer.StartSpan first calls Tracer.CreateSpanContext and then later calls this method TraceContext.AddSpan, and Tracer.CreateSpanContext actually does this exact same thing, so if at this point we see that `_samplingPriority == null` then this is condition cannot be true

Reply:
What Zach said. In other words:

- if `span` is the local root span (i.e. the first span added to the trace context), then
  - if `span` was created from a propagated context and contains the sampling decision made upstream, and we have already set `TraceContext.SamplingPriority` before hitting this code
  - if `span` was _not_ created from a propagated context (i.e. it is the real root, parentId == 0), or if for some reason the propagated context didn't contain a sampling priority, then we make our own sampling decision here (we can't do it earlier because the sampling decision is a function of service name etc so we have to wait until we have the first span)